README
WebDyne::Request::FastCGI is part of the WebDyne module. It provides a
FastCGI wrapper so WebDyne will work with FastCGI servers such as Lighttpd. 
It should also work with other FastCGI servers such as Zeus, and works with
the Windows IIS FastCGI module.

WebDyne::Request::FastCGI will only work within WebDyne and will not
function as a stand-alone module.


ChangeLog
commit 68a4d4a1b2947590756e8372013db92658c98bcd
Author: Andrew Speer <andrew@webdyne.org>
Date:   Wed Jul 13 17:22:03 2011 +0930

    Tag: WebDyne-Request-FastCGI_1.019
    
    Modify Makefile PREREQ section to be more compatible with older version of CPAN

commit c0222d84a268df1e7d265ac49f892e28258f4ee5
Author: Andrew Speer <andrew@webdyne.org>
Date:   Mon Apr 12 12:41:29 2010 +0930

    Tag: WebDyne-Request-FastCGI_1.018
    
    change tag used for text errors from tt=>pre to preserve cr/lf
    clean mm_ver redundancy in Makefile.PL, edit README

commit 8ce00b3230d970ba9aed73c6fab979376a803172
Author: Andrew Speer <andrew@webdyne.org>
Date:   Sun Apr 11 21:36:42 2010 +0930

    Tag: WebDyne-Request-FastCGI_1.017
    
    Cleaned up wdfastcgi binary, removed extraneous ref to File::Spec and
    fixed incorrect comments

commit d61f480661a0c3ee9ba5b302796faf9897879c90
Author: Andrew Speer <andrew@webdyne.org>
Date:   Sun Apr 11 21:24:57 2010 +0930

    Tag: WebDyne-Request-FastCGI_1.016
    
    Updated to YAML 1.4

commit ff337bc063a3666d39b7e934e4a3d819e5f835a2
Author: Andrew Speer <andrew@webdyne.org>
Date:   Sun Apr 11 16:45:22 2010 +0930

    Tag: WebDyne-Request-FastCGI_1.015
    
    Corrected email address

commit 39c586857145395e44908a476a468d4ae5242823
Author: Andrew Speer <andrew@webdyne.org>
Date:   Sun Apr 11 16:23:47 2010 +0930

    Tag: WebDyne-Request-FastCGI_1.014
    
    Fix copyright and author information, update Makefile.PL to generate better META.yml

commit bb640a866b2a1960154d13a5531e5771225a5f62
Author: Andrew Speer <aspeer@copper.isolutions.com.au>
Date:   Thu Apr 8 22:09:25 2010 +0930

    Tag: WebDyne-Request-FastCGI_1.013
    
    Cleaned up ready for CPAN release, add rudimentry module load test, update documentation

commit 3c872046877882103c7e4844cb8de5b79f6755f4
Author: Andrew Speer <aspeer@silicon.isolutions.com.au>
Date:   Mon Aug 24 02:19:18 2009 +0930

    Tag: WebDyne-Request-FastCGI_1.12
    
    Cleanup after initial import into Git

commit 9e49a2e868c0573cae1a8947825b0c43d475a348
Author: Andrew Speer <aspeer@silicon.isolutions.com.au>
Date:   Sat Aug 22 15:13:06 2009 +0930

    Tag: WebDyne-Request-FastCGI_1.10



Makefile.PL
#  Makefile.PL for WebDyne::Request::FastCGI
#
package WebDyne::Request::FastCGI;


#  Compiler Pragma
#
use 5.006;
use strict qw(vars);


#  External Modules
#
use ExtUtils::MakeMaker;
use File::Spec;
use IO::File;


#  Get MakeMaker version - from Test::Simple Makefile.PL
#
my $mm_ver = $ExtUtils::MakeMaker::VERSION;
if ($mm_ver =~ /_/) { # dev version
    $mm_ver = eval $mm_ver;
    die $@ if $@;
}


#  All done, init finished
#
#------------------------------------------------------------------------------


#  Create the Makefile params
#
my $prereq_hr;
my %makefile=(

    NAME					=>  __PACKAGE__,
    VERSION_FROM			    	=>  File::Spec->catfile(

	'lib',
	split(/::/, __PACKAGE__),
	'VERSION.pm'

       ),

    PREREQ_PM					=>  do { $prereq_hr={

	'WebDyne'			        =>  0,
	'FCGI'					=>  0

       }; ($ENV{'PREREQ_PM'} eq '0') ? {} : $prereq_hr },


    CONFIGURE                               	=>  sub { {

	PREREQ_PM			    	=>  $prereq_hr

    } },

    EXE_FILES				    	=>  [

	'bin/wdfastcgi'

       ],

    'dist'					=>  {

	COMPRESS				=>  'gzip -9f',
	SUFFIX					=>  '.gz'

       },

    'depend'					=>  {

	Makefile				=>  '$(VERSION_FROM)'

       },

    'macro'					=>  {

        COPYRIGHT				=>  'Copyright (C) 2006-2010 Andrew Speer, All rights reserved',
	LICENSE					=>  'GPL: GNU General Public License'

       },

    ($] <  5.005 ? () : (
    
        AUTHOR				    =>  'Andrew Speer <andrew@webdyne.org>',
        
    )),

    ($mm_ver < 6.31 ? () : (

        LICENSE                             =>  'gpl',

    )),
    
    ($mm_ver <= 6.45 ? () : (
        
        META_MERGE => { resources => {
            
            homepage                        =>  'http://webdyne.org',
            license                         =>  'http://www.gnu.org/licenses/gpl.html'
            
        }}
    )),

   );


#  Debugging on or off ?
#
my $debug=$ENV{'WEBDYNE_DEBUG'} ? 1 : 0;


#  Get ExtUtils::MakeMaker version. see if it supports PM_FILTER. From SpamAssassin Makefile.PL
#
my $mm_has_pm_filter=$mm_ver >  5.45 || ($mm_ver == 5.45 && $] > 5.006);


#  Install if supported
#
if (!$debug && $mm_has_pm_filter) {


    #  Write filter script and assign
    #
    my $fn='pm_filter.pl';
    my $fh=IO::File->new($fn, O_WRONLY|O_CREAT|O_TRUNC) ||
	die ("unable to write filter file '$fn', $!");
    print $fh 's/debug\\(/0 && debug\\(/m; print';
    $fh->close();
    my $pm_filter="\$(PERL) -nl $fn";
    $makefile{'PM_FILTER'}=$pm_filter;

}

WriteMakefile(%makefile);

package MY;

sub constants {


    #  Add license file, other support files here
    #
    my $mm_or=shift();
    $mm_or->{'PM'}{'LICENSE'}='$(INST_LIBDIR)/$(BASEEXT)/LICENSE';
    $mm_or->SUPER::constants;

};




lib/WebDyne/Request/FastCGI.pm
#
#
#  Copyright (C) 2006-2010 Andrew Speer <andrew@webdyne.org>. All rights 
#  reserved.
#
#  This file is part of WebDyne::Request::FastCGI
#
#  WebDyne is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
#
package WebDyne::Request::FastCGI;


#  Compiler Pragma
#
use strict	qw(vars);
use vars	qw($VERSION @ISA);


#  External modules
#
use File::Spec::Unix;
use HTTP::Status qw(status_message RC_OK RC_NOT_FOUND RC_FOUND);
use URI;


#  WebDyne modules
#
use WebDyne::Request::FastCGI::Constant;
use WebDyne::Base;


#  Inheritance
#
use WebDyne::Request::Fake;
@ISA=qw(WebDyne::Request::Fake);


#  Version information
#
$VERSION='1.004';


#  Debug load
#
debug("Loading %s version $VERSION", __PACKAGE__);


#  Save local copy of environment for ref by Dir_config handler. ENV is reset for each request,
#  so must use a snapshot for simulating r->dir_config
#
my %Dir_config_env=%ENV;


#  All done. Positive return
#
1;


#==================================================================================================


sub content_type {

    my $r=shift();
    my $hr=$r->headers_out();
    @_ ? $hr->{'Content-Type'}=shift() :  $hr->{'Content-Type'};

}


sub custom_response {

    my ($r,$status)=(shift(), shift());
    while ($r->prev) { $r=$r->prev }
    debug("in custom response, status $status");
    @_ ? $r->{'custom_response'}{$status}=\@_ : $r->{'custom_response'}{$status};

}


sub filename {

    my $r=shift();
    @_ ? $r->{'filename'}=shift() : $r->{'filename'};

}


sub header_only {

    ($ENV{'REQUEST_METHOD'} eq 'HEAD') ? 1 : 0;

}


sub headers_in {

    my ($r, $header)=@_;
    my $hr=$r->{'headers_in'} ||= do {
	my @http_header=grep { /^HTTP_/ } keys %ENV;
	my %http_header=map { $_=>$ENV{$_} } @http_header;
	foreach my $k (keys %http_header) {
	    my $v=delete $http_header{$k};
	    $k=~s/^HTTP_//;
	    $k=~s/_/-/g;
	    $http_header{lc($k)}=$v;
	}
	\%http_header;
    };
    $header ? $hr->{lc($header)} : $hr;

}


sub dir_config0 {

    my ($r, $key)=@_;

    my $constant_hr=$WEBDYNE_DIR_CONFIG;
    my @location=grep {$_} 
        File::Spec::Unix->rootdir(), File::Spec::Unix->splitdir($r->location());
    debug("in dir_config looking for key $key");
    
    
    while (my $location=File::Spec::Unix->catdir(@location)) {
        debug("location $location");
        if (exists $constant_hr->{$location}) {
            return $constant_hr->{$location}{$key} 
                #|| $constant_hr->{undef()}{$key} || $Dir_config_env{$key}
        }
        elsif (exists $constant_hr->{$location.=File::Spec->rootdir()}) {
            return $constant_hr->{$location}{$key} 
        }
        pop @location;
    }


    debug("explicit location key $key not found, returning top level");
    return $constant_hr->{''}{$key} || $Dir_config_env{$key}

}

sub dir_config {

    my ($r, $key)=@_;

    my $constant_hr=$WEBDYNE_DIR_CONFIG;

    my $constant_server_hr;
    if (my $server=$Dir_config_env{'WebDyneServer'} || $ENV{'SERVER_NAME'}) {
	$constant_server_hr=$constant_hr->{$server} if exists($constant_hr->{$server})
    }

    my $location=$r->location();
    debug("in dir_config looking for key $key at location $location");

    if (exists $constant_server_hr->{$location}) {
	return $constant_server_hr->{$location}{$key} 
    }    
    elsif (exists $constant_hr->{$location}) {
	return $constant_hr->{$location}{$key} 
	#|| $constant_hr->{undef()}{$key} || $Dir_config_env{$key}
    }
    else {
	debug("explicit location key $key not found, returning top level");
	return $constant_hr->{''}{$key} || $Dir_config_env{$key}
    }

}


sub location0 {

    if ($ENV{'SCRIPT_NAME'}) {
        return (File::Spec::Unix->splitpath($ENV{'SCRIPT_NAME'}))[1];
    }
    else {
        #  APPL_MD_PATH is IIS virtual dir
        #
        return $Dir_config_env{'WebDyneLocation'} || $ENV{'APPL_MD_PATH'} || '/';
    }

}


sub location {
    
    #  Equiv to Apache::RequestUtil->location;
    #
    my $location;
    my $constant_hr=$WEBDYNE_DIR_CONFIG;
    my $constant_server_hr;
    if (my $server=$Dir_config_env{'WebDyneServer'} || $ENV{'SERVER_NAME'}) {
	$constant_server_hr=$constant_hr->{$server} if exists($constant_hr->{$server})
    }
    if ($Dir_config_env{'WebDyneLocation'} || $ENV{'APPL_MD_PATH'}) {
        #  APPL_MD_PATH is IIS virtual dir
        #
	$location=$Dir_config_env{'WebDyneLocation'} || $ENV{'APPL_MD_PATH'};
    }
    elsif ($ENV{'SCRIPT_NAME'}) {
        my $path=(File::Spec::Unix->splitpath($ENV{'SCRIPT_NAME'}))[1];
        my @location=grep {$_} 
	    File::Spec::Unix->rootdir(), File::Spec::Unix->splitdir($path);
	while ($location=File::Spec::Unix->catdir(@location)) {
	    debug("location $location");
	    last if exists ($constant_hr->{$location}) || exists ($constant_server_hr->{$location});
	    $location.=File::Spec::Unix->rootdir();
	    last if exists ($constant_hr->{$location}) || exists ($constant_server_hr->{$location});
	    pop @location;
	}
    }
    else {
	$location = File::Spec::Unix->rootdir();
    }
    return $location;

}


sub log_error {

    shift(); warn(@_) if $WEBDYNE_FASTCGI_WARN_ON_ERROR;

}


sub lookup_file {

    my ($r, $fn)=@_;
    my $r_child;
    if ($fn=~/\.html$/) {


	#  Static file
	#
	require WebDyne::Request::FastCGI::Static;
	$r_child=WebDyne::Request::FastCGI::Static->new( filename=>$fn, prev=>$r ) ||
	    return err();

    }
    else {


	#  Subrequest
	#
	$r_child=ref($r)->new( filename=> $fn, prev=>$r ) || return err();

    }

    #  Return child
    #
    return $r_child;

}


sub lookup_uri {

    my ($r, $uri)=@_;
    ref($r)->new( uri=>$uri, prev=>$r ) || return err();

}


sub new {

    my ($class, %r)=@_;
    unless ($r{'filename'}) {

	my $fn;
	unless (($fn=$ENV{'SCRIPT_FILENAME'}) &&  !$r{'uri'}) {

	    if (my $dn=$ENV{'DOCUMENT_ROOT'}) {
	        my $uri=$r{'uri'} || $ENV{'REQUEST_URI'};
		if (my $location=$class->location()) {
		    $uri=~s/^\Q$location\E//;
		}
		my $uri_or=URI->new($uri);
		$fn=File::Spec->catfile($dn, $uri_or->path());
	    }

	    elsif ($fn=$ENV{'PATH_TRANSLATED'}) {
		#  Feel free to let me know a better way under IIS/FastCGI ..
		my $script_fn=(File::Spec::Unix->splitpath($ENV{'SCRIPT_NAME'}))[2];
		$fn=~s/\Q$script_fn\E.*/$script_fn/;
	    }

	}

	$r{'filename'}=$fn || do {
	    my $env=join("\n", map {"$_=$ENV{$_}"} keys %ENV);
	    return err("unable to determine filename for request from environment: $env")
	};
	
    }

    bless \%r, $class;

}


sub redirect {

    my ($r, $location)=@_;
    CORE::print sprintf("Status: %s\r\n", RC_FOUND);
    CORE::print "Location: $location\r\n";
    $r->send_http_header;
    return RC_FOUND;

}


sub run {

    my ($r, $self)=@_;
    if (-f $r->{'filename'}) {
	return ref($self)->handler($r);
    }
    else {
	debug("file not found !");
	$r->status(RC_NOT_FOUND);
	$r->send_error_message;
	RC_NOT_FOUND;
    }

}


sub set_handlers {

    #  No-op

}


sub send_error_response {

    my $r=shift();
    my $status=$r->status();
    CORE::print "Status: $status\r\n";
    $r->send_http_header;
    debug("in send error response, status $status");
    if (my $message_ar=$r->custom_response($status)) {

	#  We have a custom response - send it
	#
	$r->print(@{$message_ar});

    }
    else {

	#  Create an generic error message
	#
	$r->print($r->err_html(
	    $status,
	    status_message($status)
	   ));
    }
}


sub err_html {

    my ($r,$status,$message)=@_;
    require CGI;
    my $error;
    my @message=(
	CGI->start_html($error=sprintf("%s Error $status", __PACKAGE__)),
	CGI->h1($error),
	CGI->hr(),
	CGI->em(status_message($status)), CGI->br(), CGI->br(),
	CGI->pre(
	    sprintf("The requested URI '%s' generated error:\n\n$message", $r->uri)
	),
	CGI->end_html()
       );
    return join(undef, @message);

}


sub send_http_header {

    my $r=shift();
    while(my ($h,$v)=each %{$r->headers_out()}) {
	CORE::print "$h: $v\r\n";
    }
    CORE::print "\r\n";

}


sub uri {

    my $r=shift();
    @_ ? $r->{'uri'}=shift() : $r->{'uri'} || $ENV{'REQUEST_URI'}

}


sub protocol {
    
    $ENV{'SERVER_PROTOCOL'}
}


sub env {

    return \%Dir_config_env;
    
}

__END__

=head1 Name

WebDyne::Request::FastCGI - FastCGI responder that handles requests for WebDyne pages.

=head1 Description

WebDyne::Request::FastCGI is a FastCGI responder that handles requests for WebDyne pages made to FastCGI based server
such as Lighttd, LiteSpeed etc.

=head1 Documentation

Information on configuration and usage is availeble from the WebDyne site, http://webdyne.org/ - or from a snapshot of
current documentation in PDF format available in the WebDyne source /doc directory.

=head1 Copyright and License

WebDyne::Request::FastCGI is Copyright (C) 2006-2010 Andrew Speer.  WebDyne::Request::FastCGI is dual licensed. It is
released as free software released under the Gnu Public License (GPL), but is also available for commercial use under
a proprietary license - please contact the author for further information.

WebDyne::Request::FastCGI is written in Perl and uses modules from CPAN (the Comprehensive Perl Archive Network).
CPAN modules are Copyright (C) the owner/author, and are available in source from CPAN directly. All CPAN modules used
are covered by the Perl Artistic License.

=head1 Author

Andrew Speer, andrew@webdyne.org

=head1 Bugs

Please report any bugs or feature requests to "bug-webdyne-request-fastcgi at rt.cpan.org", or via
http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WebDyne-Request-FastCGI



lib/WebDyne/Request/FastCGI/Static.pm
#
#
#  Copyright (c) 2003 Andrew W. Speer <andrew.speer@isolutions.com.au>. All rights 
#  reserved.
#
#  This file is part of WebDyne
#
#  WebDyne is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
#
#  $Id: WebDyne_Request_FastCGI_Static.pm,v 1.1.1.1 2006/06/12 07:55:33 aspeer Exp $
#
package WebDyne::Request::FastCGI::Static;


#  Compiler Pragma
#
use strict	qw(vars);
use vars	qw($VERSION $AUTOLOAD @ISA);


#  External modules
#
use HTTP::Status (qw(RC_INTERNAL_SERVER_ERROR RC_NOT_FOUND));
use IO::File;
use WebDyne::Base;


#  Inheritance
#
use WebDyne::Request::FastCGI;
@ISA=qw(WebDyne::Request::FastCGI);


#  Version information
#
$VERSION='1.002';


#  Debug load
#
debug("Loading %s version $VERSION", __PACKAGE__);


#  All done. Positive return
#
1;


#==================================================================================================


sub run {

    my ($r, $self)=@_;
    my $fn=$r->filename();
    if (!-f $fn) {
	warn("file '$fn' not found");
	return $r->status(RC_NOT_FOUND);
    }
    elsif (my $fh=IO::File->new($fn, O_RDONLY)) {
	my $hr=$r->headers_out();
	my $size=(stat($fn))[7];
	$hr->{'Content-Length'}=$size;
	$r->send_http_header();
	while (<$fh>) { $r->print($_) }
	$fh->close();
	return &Apache::OK
    }
    else {
	warn("unable to open file '$fn', $!");
	return $r->status(RC_INTERNAL_SERVER_ERROR);
    }

}


lib/WebDyne/Request/FastCGI/VERSION.pm
#
#
#  Copyright (C) 2006-2010 Andrew Speer <andrew@webdyne.org>. All rights
#  reserved.
#
#  This file is part of WebDyne::Request::FastCGI.
#
#  WebDyne::Request::FastCGI is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
#
package WebDyne::Request::FastCGI::VERSION;
$VERSION='1.019';


lib/WebDyne/Request/FastCGI/Constant.pm
#
#
#  Copyright (c) 2003 Andrew W. Speer <andrew.speer@isolutions.com.au>. All rights 
#  reserved.
#
#  This file is part of WebDyne.
#
#  WebDyne::Session is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
#
#  $Id: WebDyne_Request_FastCGI_Constant.pm,v 1.3 2007/10/28 15:02:23 aspeer Exp $


#  Constants
#
package WebDyne::Request::FastCGI::Constant;
use strict qw(vars);
use vars qw($VERSION @ISA %EXPORT_TAGS @EXPORT_OK @EXPORT %Constant);
no warnings qw(uninitialized);
local $^W=0;


#  Version information. Must be all on one line
#
$VERSION='1.002';


#  The guts
#
%Constant = (


    #  A placeholder for per-location WebDyne settings, e.g., WebDyneHandler, etc.
    #
    WEBDYNE_DIR_CONFIG		    =>  undef,


    #  Warn on errors ? Some FastCGI implementations write to error logs (good), others send to client (bad)
    #
    WEBDYNE_FASTCGI_WARN_ON_ERROR   =>  undef,


    #  Max requests on socket if not given see FCGI::OpenSocket
    #
    WEBDYNE_FASTCGI_BACKLOG	    =>  4,


);


#  Export constants to namespace, place in export tags
#
require Exporter;
require WebDyne::Constant;
@ISA=qw(Exporter WebDyne::Constant);
+__PACKAGE__->local_constant_load(\%Constant);
map { ${$_}=$Constant{$_} } keys %Constant;
@EXPORT=map { '$'.$_ } keys %Constant;
@EXPORT_OK=@EXPORT;
%EXPORT_TAGS=(all => [@EXPORT_OK]);
$_=\%Constant;


lib/WebDyne/Request/FastCGI/Run.pm
#
#
#  Copyright (c) 2003 Andrew W. Speer <andrew.speer@isolutions.com.au>. All rights 
#  reserved.
#
#  This file is part of WebDyne.
#
#  WebDyne is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
#
#  $Id: WebDyne_Request_FastCGI_Run.pm,v 1.2 2007/10/28 15:02:23 aspeer Exp $
#
package WebDyne::Request::FastCGI::Run;


#  Compiler Pragma
#
use strict	qw(vars);
use vars	qw($VERSION);


#  External modules
#
use WebDyne;
use WebDyne::Base;
use WebDyne::Request::FastCGI;
use WebDyne::Request::FastCGI::Constant;
use FCGI;
use CGI;
use HTTP::Status qw(RC_INTERNAL_SERVER_ERROR RC_NOT_FOUND);


#  Set handler as constant. No longer do this - load hander on per request basis on
#  information from dir_config.
#
#use constant WebDyneHandler=>$ENV{'WebDyneHandler'} ? $ENV{'WebDyneHandler'} : 'WebDyne';


#  Load up whichever handler we are using
#
#my $handler=WebDyneHandler;
#eval ("require $handler") || die("unable to load handler $handler, $@");


#  Version information
#
$VERSION='1.002';


#  Signal handlers (seem broken with lighttp, ignore for now
#
#$SIG{'USR1'}=sub { exit 0 };
#$SIG{'TERM'}=sub { exit 0 };
#$SIG{'PIPE'}='IGNORE';


#  Sig handler
#
#sub sig {
#    exit(0) if !$busy;
#    $exit++;
#}


#  CGI needs to be hacked under some versions of FastCGI so path_info returns correct string, otherwise
#  CGI->self_url does not return correct value, then forms with explicit action attr stop working
#
my $path_info_cr=\&CGI::path_info || die('unable to get CGI code ref for path_info');
*CGI::path_info=sub {
    my $path_info=$path_info_cr->(@_);
    $path_info=~s/^\Q$ENV{'SCRIPT_NAME'}\E//;
    $path_info;
};



#  Debug
#
#debug("wdfastcgi handler for pid $$ is $handler, env %s", &Data::Dumper::Dumper(\%ENV));


#  All done. Positive return
#
1;


#==================================================================================================


#  Start endless loop
#
sub start {


    #  Can optionally be passed a socket file name at start
    #
    my ($class, $socket_fn, $backlog)=@_;
    my $socket_fh=$socket_fn && do {
	$backlog ||=  WEBDYNE_FASTCGI_BACKLOG;
	FCGI::OpenSocket($socket_fn, $backlog) ||
	    die ("could not open socket $socket_fn, $!");
    };
    
    
    #  Workaround for perlbug 73672
    #
    my %env=%ENV;
    *ENV=\*env;


    #  Cache handler for a location
    #
    my ($handler, %handler);


    #  Get FastCGI request object
    #
    my $fcgi_r=$socket_fh ?
	FCGI::Request(\*STDIN, \*STDOUT, \*STDERR, \%ENV, $socket_fh) : FCGI::Request();
    my ($busy, $exit);
    
    
    while($busy=($fcgi_r->Accept() >= 0)) {


	#  Create new FastCGI Request object, will pull filename from
	#  environment. goto only invoked if no object created, which
	#  should never happen !
	#
    	my $r=WebDyne::Request::FastCGI->new() ||
	    goto RENDER_ERROR;
	    
	    
        #  Get handler
        #
        unless ($handler=$handler{my $location=$r->location()}) {
            my $handler_package=
                $r->dir_config('WebDyneHandler') || $ENV{'WebDyneHandler'};
            if ($handler_package) {
                local $SIG{'__DIE__'};
                unless (eval("require $handler_package")) {
                    #  Didn't load - let Webdyne handle the error.
                    $handler='WebDyne';
                }
                else {
                    $handler=$handler{$location}=$handler_package;
                }
            }
            else {
                $handler=$handler{$location}='WebDyne';
            }
        }


	#  Call handler and evaluate results
	#
	my $status=eval { $handler->handler($r) } if $handler;
	if (!defined($status)) {
	    if (($status=$r->status) ne RC_INTERNAL_SERVER_ERROR) {
		my $error=errdump() || $@; errclr();
		$r->custom_response(
		    $r->status($status=RC_INTERNAL_SERVER_ERROR),
		    $r->err_html($status, $error)
		);
	    }
	}
	elsif (($status < 0) && !(-f (my $fn=$r->filename()))) {
	    $r->status($status=RC_NOT_FOUND);
	    warn("file $fn not found") if $WEBDYNE_FASTCGI_WARN_ON_ERROR;
	}
	elsif ($status < 0) {
	    $r->custom_response(
		$r->status($status=RC_INTERNAL_SERVER_ERROR),
		$r->err_html($status, "Unexpected return ($status) from handler $handler")
	       );
	}
	debug("handler status was $status");


	#  If status ne Apache::OK invoke error handler
	#
	$status && $r->send_error_response();


	#  Done, call destoy manually so cleanup_handlers run
	#
	$r->DESTROY;


	#  Signal handlers broken, these never get checked
	#$busy=0;
	#last if $exit;
	next;


	#  Error handler, exceptional cases only (usually handled by
	#  WebDyne handler).
	#
	RENDER_ERROR:
	require CGI;
	my $error=errdump() || 'unknown error'; errclr();
	CORE::print 
	    sprintf("Status: %s\r\n", RC_INTERNAL_SERVER_ERROR),
	    "Content-Type: text\html\r\n\r\n",
		WebDyne::Request::FastCGI->err_html(RC_INTERNAL_SERVER_ERROR, CGI->pre($error));

    };

    #  All done, quit. Commented out due to broken sig handlers
    #
    #$fcgi_r->Finish();
    #exit(0);

}

